test/integration/associations/belongs-to.test.js

```diff
@@ -836,3 +836,76 @@ describe(Support.getTestDialectTeaser('BelongsTo'), function() {
     });
   });
 });
+
+describe('Association', function() {
+  it('should set foreignKey on foreign table', function () {
+    const Mail = this.sequelize.define('mail', {});
+    const Entry = this.sequelize.define('entry', {});
+    const User = this.sequelize.define('user', {});
+    Entry.belongsTo(User, { as: 'owner', foreignKey: { name: 'ownerId', allowNull: false } });
+    Entry.belongsTo(Mail, {
+      as: 'mail',
+      foreignKey: {
+        name: 'mailId',
+        allowNull: false
+      }
+    });
+    Mail.belongsToMany(User, {
+      as: 'recipients',
+      through: 'MailRecipients',
+      otherKey: {
+        name: 'recipientId',
+        allowNull: false
+      },
+      foreignKey: {
+        name: 'mailId',
+        allowNull: false
+      }
+    });
+    Mail.hasMany(Entry, {
+      as: 'entries',
+      foreignKey: {
+        name: 'mailId',
+        allowNull: false
+      }
+    });
+    User.hasMany(Entry, {
+      as: 'entries',
+      foreignKey: {
+        name: 'ownerId',
+        allowNull: false
+      }
+    });
+    return this.sequelize.sync({ force: true })
+      .then(() => User.create({}))
+      .then(() => Mail.create({}))
+      .then(mail =>
+        Entry.create({ mailId: mail.id, ownerId: 1 })
+          .then(() => Entry.create({ mailId: mail.id, ownerId: 1 }))
+          // set recipients
+          .then(() => mail.setRecipients([1]))
+      )
+      .then(() => Entry.findAndCount({
+        offset: 0,
+        limit: 10,
+        order: [['id', 'DESC']],
+        include: [
+          {
+            association: Entry.associations.mail,
+            include: [
+              {
+                association: Mail.associations.recipients,
+                through: {
+                  where: {
+                    recipientId: 1
+                  }
+                },
+                required: true
+              }
+            ],
+            required: true
+          }
+        ]
+      }));
```

Comment:
Would love an expectation here, if you're simply testing that SQL is generated correctly we'd prefer a unit test for execution speed.